pandas/core/indexing.py

```diff
@@ -1043,12 +1051,17 @@ def _getitem_iterable(self, key, axis=0):
             return self.obj.take(inds, axis=axis, convert=False)
         else:
             if isinstance(key, Index):
-                # want Index objects to pass through untouched
-                keyarr = key
+                if isinstance(labels, ABCUInt64Index) and key.is_integer():
+                    keyarr = key.astype(np.uint64)
```

Comment:
this should be handled using ``_convert_index_indexer`` (e.g. like ``_convert_list_indexer``), a method on the ``Index`` object itself (and for all indexes will basically just return, except for UInt64 you can do the astyping).